### PR TITLE
gRPC probes support (KEP 2727)

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
https://github.com/kubernetes/enhancements/issues/2727

/sig node
/hold

### CODE REVIEW

**Comment by jlbutler** [comment]
@SergeyKanzhelev Reminder that tomorrow Nov 23rd is the deadline for 1.23 Docs Ready for Review. Please push your docs content for this enhancement and remove the work-in-progress label when ready for review.

We have a tight turnaround to getting docs reviewed and approved for the release, and we’re hoping to have all docs in for review by EOB tomorrow. Thank you!

**Comment by sftim** [comment]
Hi @SergeyKanzhelev 

Would you be willing to revise this to be ready for review?

**Comment by SergeyKanzhelev** [comment]
> Hi @SergeyKanzhelev
> 
> Would you be willing to revise this to be ready for review?

I'm still working on validating yaml. some issues setting up 1.23 environment. But facts-wise, ready for review

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 1]
```diff
@@ -220,11 +220,53 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+{{< codenew file="pods/probe/content/en/examples/pods/probe/grpc-liveness.yaml">}}
+
+For the gRPC probe `port` have to be configured. If the health endpoint is configured
+on non-default service, `service` must be configured.
+
+{{< note >}}
+Unlike http and tcp probes, named ports cannot be used and custom host cannot be configured.
+{{< /note >}}
+
+Configuration problems - incorrect port and service, or not implemented health
```
Comment (jlbutler) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:239]: suggestion:  "Configuration problems (e.g. incorrect port and service, unimplemented health checking protocol) are considered a probe failure, similar to HTTP and TCP probes."

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 2]
```diff
@@ -220,11 +220,53 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+checking protocol are considered a probe failure similar to http and tcp probes.
+
+Before Kubernetes 1.23, gRPC health probes were often implemented using [grpc-health-probe](https://github.com/grpc-ecosystem/grpc-health-probe/),
+like described in the blog post [Health checking gRPC servers on Kubernetes](/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/).
```
Comment (jlbutler) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:243]: nit "as described"

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 3]
```diff
@@ -220,11 +220,53 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+The built-in gRPC probes behavior is similar to one implemented by grpc-health-probe.
+When migrating from grpc-health-probe to built-in probes, remember the following differences:
+
```
Comment (jlbutler) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:246]: nit, capitalize first letter in each of the bulleted items

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 4]
```diff
@@ -220,11 +220,53 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+- built-in probes will run against pod IP, unlike grpc-health-probe that often runs against `127.0.0.1`.
+  Be sure to configure your gRPC endpoint to listen for pod IP address.
+- built-in probes do not currently support any authentication parameters (like `-tls`).
+- there is no error codes in built-in probes. All errors are considered a probe failures.
+
+In the example below, etcd pod is configured to use gRPC liveness probe.
+To try the gRPC liveness check, create a Pod:
```
Comment (jlbutler) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:253]: maybe flip these two sentences for clarity? e.g. "To try the gRPC liveness check, create a Pod. In this example, the etcd pod is configured to use the liveness probe."

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 5]
```diff
@@ -220,11 +220,53 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+
+```shell
+kubectl apply -f https://k8s.io/examples/pods/probe/content/en/examples/pods/probe/grpc-liveness.yaml
+```
+
+After 15 seconds, view Pod events to verify that liveness probes run:
```
Comment (jlbutler) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:259]: "...to verify that the liveness probes have run"

**Review by jlbutler — CHANGES_REQUESTED** [review]
Thanks for getting this in! Suggested some wording/nits.

File: content/en/blog/_posts/2018-10-01-health-checking-grpc.md [hunk 6]
```diff
@@ -4,6 +4,9 @@ title:  'Health checking gRPC servers on Kubernetes'
 date: 2018-10-01
 ---
 
+_Since this article publication, in Kubernetes 1.23, built-in gRPC probes
```
[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply content/en/blog/_posts/2018-10-01-health-checking-grpc.md:7]: Example: https://kubernetes.io/blog/2016/11/visualize-kubelet-performance-with-node-dashboard/

I think it is a big benefit for end users to add those notes and save some time for readers.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 7]
```diff
@@ -573,10 +574,10 @@ Each feature gate is designed for enabling/disabling a specific feature:
   extended tokens by starting `kube-apiserver` with flag `--service-account-extend-token-expiration=false`.
   Check [Bound Service Account Tokens](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/1205-bound-service-account-tokens/README.md)
    for more details.
-- `ControllerManagerLeaderMigration`: Enables Leader Migration for 
-  [kube-controller-manager](/docs/tasks/administer-cluster/controller-manager-leader-migration/#initial-leader-migration-configuration) and 
-  [cloud-controller-manager](/docs/tasks/administer-cluster/controller-manager-leader-migration/#deploy-cloud-controller-manager) which allows a cluster operator to live migrate 
-  controllers from the kube-controller-manager into an external controller-manager 
+- `ControllerManagerLeaderMigration`: Enables Leader Migration for
```
[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:577]: this is my auto-formatter for md files. lmk if this is critical and I will fix it back

[the remark this answers is not in the document]
Comment (jlbutler) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:577]: No, fine by me.

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 8]
```diff
@@ -220,11 +220,54 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+
+Before Kubernetes 1.23, gRPC health probes were often implemented using [grpc-health-probe](https://github.com/grpc-ecosystem/grpc-health-probe/),
+as described in the blog post [Health checking gRPC servers on Kubernetes](/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/).
+The built-in gRPC probes behavior is similar to one implemented by grpc-health-probe.
+When migrating from grpc-health-probe to built-in probes, remember the following differences:
+
+- Built-in probes will run against pod IP, unlike grpc-health-probe that often runs against `127.0.0.1`.
+  Be sure to configure your gRPC endpoint to listen for pod IP address.
+- Built-in probes do not currently support any authentication parameters (like `-tls`).
+- There is no error codes in built-in probes. All errors are considered a probe failures.
```
Comment (mrunalp) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:250]: There are no error codes in built-in probes. All errors are considered as probe failures.

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 9]
```diff
@@ -220,11 +220,54 @@ After 15 seconds, view Pod events to verify that liveness probes:
...
+
+To try the gRPC liveness check, create a Pod using command below.
```
Comment (mrunalp) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:252]: using the command

**Comment by mrunalp** [comment]
I left a couple of nits. Otherwise
/lgtm

**Comment by SergeyKanzhelev** [comment]
addressed @mrunalp's comments. Also added a small note. This is ready for merge

File: content/en/blog/_posts/2018-10-01-health-checking-grpc.md [hunk 10]
```diff
@@ -4,6 +4,8 @@ title:  'Health checking gRPC servers on Kubernetes'
 date: 2018-10-01
 ---
 
+_Built-in gRPC probes were introduced in Kubernetes 1.23. To learn more, see [Configure Liveness, Readiness and Startup Probes](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe)._
```
Comment (sftim) [inline content/en/blog/_posts/2018-10-01-health-checking-grpc.md:7]: Nice tweak.

**Pull request merged**